### PR TITLE
Met API

### PR DESCRIPTION
**Pull request by FatimahAreola**
In this PR:
- Return details from the MET API for a set of pre-determined pieces of art 
- Created Folder/Document for us to keep track of testing guidelines

Things to note: 
- I've added json to the requirements.txt document, it is important for anyone trying to run the code to be aware of this

### CODE REVIEW

File: app/app.py [hunk 1]
```diff
@@ -9,5 +10,36 @@ def hello():
...
+    #Define index, and Cardset dictionary
+    idx=0
+    cardSet={}  
+#This for loop goes through all IDs provided in the list of art pieces chosen and pulls the required information about them from the MET's API
+    for i in artObjectIDs:
+        apiResponse = requests.get("https://collectionapi.metmuseum.org/public/collection/v1/objects/" + str(i))
+        artDetails = apiResponse.json()
+#We add another object to the Cardset dictionary for each art piece, containing info on Name, ObjectID, URL and status
+        cardSet[idx]={
+        'cardName': "Card" +str(idx+1),
```
Comment (abcotter) [inline app/app.py]: for usability to the frontend can this be just cardId
```suggestion
        'cardId': idx+1,
```

File: app/app.py [hunk 2]
```diff
@@ -9,5 +10,36 @@ def hello():
     return 'Hello This is the Flask App'
 
 
+
+""" The following code issues a call to the MET Api, and returns a json object, 
+with the details of 8 pre-determined works of art so that they can be formatted for gameplay.
+ """
+
+@app.route('/api/MetAPI')
```
Comment (abcotter) [inline app/app.py]: The execution time for this API call averaged around 4.35 seconds for me ... what what the performance for you? This is really slow, I'm wondering if the call to the MET API is slow? I would run some performance measures (time breakpoints) to try and isolate whats slowing you down and can it be adjusted to be better.

**Review by abcotter — CHANGES_REQUESTED** [review]
Please look into what is causing the API call to take so long

**Comment by abcotter** [comment]
@FatimahAreola I think we should explore running the API calls to the MET asyncronously or all at once? running them independently is probably taking longer than needed.

**Comment by FatimahAreola** [comment]
@abcotter Your suggestion makes sense, I also noticed a lag in the API calls. I do think it is worth noting that the reason I chose not to run the API calls all at once was to make sure the code was still usable when we move forward into future sprints. Down the pipeline we are looking at variable numbers of cards and pieces of art returned - as well as writing methods to determine which pieces of art to use, so making the API call a bit more dynamic was important to me. Do you think there is an effective way to both make the call more dynamic _and_ run all calls at the same time?

**Comment by abcotter** [comment]
@FatimahAreola  I agree keeping our code dynamic is important for when we introduce varying sized game boards. What if you have a helper function that takes does the call to the MET API? also we could have the API take in a parameter that repersents the game size (i.e we could pass in the number of unique cared we need boardsize/2) and then your api can pull that number of objects

I though the met api could take an array of objectIds? can't we pass them all in? you could also explore calling the api asynchronously so that they are almost lazy loading while you format the result

**Comment by FatimahAreola** [comment]
> I though the met api could take an array of objectIds? can't we pass them all in? you could also explore calling the api asynchronously so that they are almost lazy loading while you format the result

No, the MET API can't **take** an array of objectIDs, I checked. You are thinking of a different functionality. 

>hat if you have a helper function that takes does the call to the MET API? also we could have the API take in a parameter that repersents the game size (i.e we could pass in the number of unique cared we need boardsize/2) and then your api can pull that number of objects
I spent some time playing with helper functions while writing this code and found complications with passing JSON objects/dictionaries back and forth. I don't think the things you are mentioning are in scope for this sprint - happy to bring this up at our retro and factor them into future sprints. In the meantime I will continue experimenting with how this can be accomplished.

Thanks for your feedback!

**Review by CarlaLeal — COMMENTED** [review]
This looks good! I looked at the MET API and played around with it. I don't see a way to send multiple ids at once. Another very valid way around this is to use thread pooling for this issue. I think whether or not its out of scope depends on how bad the lag is on the user end? Does it make the game unusable?

One other thing can you please merge in main

**Comment by m2qaderi** [comment]
I think this looks great Fatimah. I agree with Abi's comment about the lag and Carla's suggestion sounds good but I think we should table this for now and all discuss possible solutions at our next meeting

**Review by m2qaderi — APPROVED** [review]

**Comment by FatimahAreola** [comment]
> thread pooling
Thread pooling sounds like a great suggestion! I am not familiar with it in practice so I'm happy to start researching it as a solution. The game appears playable on my end but I have requested a re-review from Abi to verify on her end.

**Comment by abcotter** [comment]
@FatimahAreola lets make sure we add a ticket for next sprint to research and implement the thread pooling Carla mentioned.

**Review by abcotter — APPROVED** [review]

**Pull request merged**